### ISSUE TITLE
domEvent: timeout waiting for node 

### ISSUE THREAD

**Issue by rjeczalik** [chromedp/chromedp#75]
This is more a question than a bug report.

When a dom event listener is about to wait for a node and the node gets concurrently removed (or its parent), WaitNode fails with timeout. The problem here is when that happens, the whole action handlers stale for 10 seconds. At least this is what I think happens. The behaviour is random, if WaitNode finishes before it's removed, then everything works as expected (I think also Go scheduler contributes to the racy behaviour).

Debug log for timeout scenario:

```
scrap.go:91: -> {"method":"DOM.childNodeInserted","params":{"parentNodeId":7,"previousNodeId":8,"node":{"nodeId":9,"backendNodeId":9,"nodeType":1,"nodeName":"DIV","localName":"div","nodeValue":"","childNodeCount":0,"attributes":[]}}}
scrap.go:91: -> {"method":"DOM.attributeModified","params":{"nodeId":9,"name":"id","value":"someotherid"}}
scrap.go:91: -> {"method":"DOM.childNodeInserted","params":{"parentNodeId":7,"previousNodeId":0,"node":{"nodeId":11,"backendNodeId":11,"nodeType":1,"nodeName":"BODY","localName":"body","nodeValue":"","childNodeCount":1,"attributes":["style","visibility: hidden; width: 0px; height: 0px; border: 0px; margin: 0px; background: none;"]}}}
scrap.go:91: -> {"method":"DOM.childNodeCountUpdated","params":{"nodeId":11,"childNodeCount":0}}
scrap.go:91: -> {"method":"DOM.childNodeRemoved","params":{"parentNodeId":7,"nodeId":9}}
scrap.go:91: -> {"method":"DOM.childNodeRemoved","params":{"parentNodeId":7,"nodeId":11}}
scrap.go:91: could not perform (attributeModified) operation on node 9 (wait node): timeout waiting for node `9`
scrap.go:91: could not perform (childNodeCountUpdated) operation on node 11 (wait node): timeout waiting for node `11`
```

I was thinking about fixing this (the 10s stales are really bugging me) by introducing some kind of context cancellation / listening channel, so we break out of waiting loop in WaitNode when the node we're waiting for gets removed.

Wdyt?

**Comment by pwaller** [comment]
I'm seeing the same behaviour. At random, if a node is removed before it is waited on, we get a 10s hang.

I have tried @rjeczalik's patch, which seems to ameliorate the problem It does give an error which I think is not useful:

```
2017/06/16 11:46:24 error: could not perform (attributeModified) operation on node 10 (wait node): node `10` was removed in the meantime
```

"I tried to update a thing which doesn't exist" isn't a useful thing to know, IMO.

More generally, rather than having [spin-wait loops](https://github.com/knq/chromedp/blob/74075d513f49e85e038579860ddf58b60d2ac96a/handler.go#L487-L518), why not instead have a subscription mechanism enabling the waiter to listen on a channel to receive  updates?

Actually, thinking at an ever higher level, is the concurrency helping here? I presume`*` the stream of events that the chrome sends is logically consistent (e.g, `node added, modified, removed` - rather than `add, remove, modify` or `modify, add, remove` which is causing the hang) - it's just that from the Go side, the events are handled asynchronously, so perhaps out of order. I'm interested to know, does this out-of-order handling on the Go side provide any benefits, or is it opening the project up to races like this?

`*` If anyone could help me with this presumption - I can't find the documentation of the protocol and I might be very wrong here in which case the above paragraph is moot.

By the way - I think `modify, add, remove` is an interesting case to consider. If `add` and `remove` happen in quick succession, then you'll still be open to this race, even if you have a mechanism to notify the `modify` waiter of the `add`, because the `modify` waiter might wake up after the removal has already completed. This I think shows that somehow, events for a given node need to be arranged in a pipeline so that any work needed for a node takes place in a sensible order. You don't want to tell the waiter "Hey, here's this `*Node`" only for it to turn out to be invalid already.

Postscript: I tried out my idea and made event handling synchronous [over here](https://github.com/pwaller/chromedp/commit/0a4d2470d6b633171cfe9d00a270bffcacc43ce8). It's a fairly minimal change which seems to fix the bug. No more spin waiting necessary. Not sure it works in all situations but it certainly seems to work well in my testing and fix the problem reported in this issue.

**Comment by pwaller** [comment]
If the [patch I made to make event handling synchronous](https://github.com/pwaller/chromedp/commit/0a4d2470d6b633171cfe9d00a270bffcacc43ce8) is deemed a good starting point, I can submit at as a pull request. Let me know. I think there could be a lot more work to follow on from that to remove now possibly unnecessary locking and asynchronous stuff. @rjeczalik I would be interested to know if it fixes your problem, and also to see if my above patch works in more strenuous cases, because I have only tested with a fairly simple case so far. I can't quite get the tests working yet either - I get `chromedp_test.go:17: could not allocate from pool: fork/exec : no such file or directory`.

**Comment by pwaller** [comment]
I hit an interesting problem with my patch and frames, in particular.

The way frames are handled seems to be a bit problematic, because both `EventFrameAttached` and `EventFrameStartedLoading` try to set properties on a `*cdf.Frame` which doesn't exist until `EventFrameNavigated` is received at some point later. It "works" at the moment in principle because each event is handled in a separate goroutine, so they can just wait until the frame arrives (spooky communication to the future!). The thing is, the only property it sets is the frame state - to `FrameAttached` and `FrameLoading`. But by the time the frame actually comes into existence the state is surely stale.

In addition, since the two of them are sitting in a sleep-spin loop, they can wake up in an arbitrary order, resulting in an arbitrary state being set (either one can win!). Is this state used for anything anyway? It seems weird that chrome is telling us stuff about frames it hasn't even sent yet.

I've also observed that when doing rapid navigations, I can end up in a situation where `EventFrameNavigated` never occurs, so the frame never pops into existence. This results in all the other events (`EventFrameStoppedLoading`, etc) sitting there waiting for a frame which won't appear.

---

I encountered a reason for the asynchronous-per-event architecture. The event queue and the "results" queue are both filled [from the same loop](https://github.com/knq/chromedp/blob/74075d513f49e85e038579860ddf58b60d2ac96a/handler.go#L158-L162). This is natural because their content both comes from one network connection. But this makes it possible to deadlock with this sequence if events are handled synchronously:

1. Receive `EventDocumentUpdated`.
2. This results in `documentUpdated` being called.
3. [`documentUpdated calls dom.GetDocument()`](https://github.com/knq/chromedp/blob/74075d513f49e85e038579860ddf58b60d2ac96a/handler.go#L288), the result of which would be received over the network connection and written into [`qres`](https://github.com/knq/chromedp/blob/74075d513f49e85e038579860ddf58b60d2ac96a/handler.go#L162). However, if multiple events come in at once, this can't happen as [`qevents <- msg`](https://github.com/knq/chromedp/blob/74075d513f49e85e038579860ddf58b60d2ac96a/handler.go#L159) may be blocked, since the event handler is processing `EventDocumentUpdated`.

The solution to this seems to be to make `qevents` have a large enough buffer that you can never receive enough events to block. This feels a bit risky, but I don't easily see another way. If I was doing this I'd make it error/panic/explode if the `qevents <- msg` ever blocks, using a select.

I've updated [my branch](https://github.com/pwaller/chromedp/tree/process-events-synchronously) with the ideas within. It seems to work pretty well in my testing so far.

**Comment by rjeczalik** [comment]
@pwaller Thanks for fixing this, your second patch works like a charm so far!

**Comment by rjeczalik** [comment]
For the record I sometimes see timeouts like this:

```
could not perform (setChildNodes) operation on node 0 (wait node): timeout waiting for node `0`
```

So I [early fail](https://github.com/rjeczalik/chromedp/commit/1d3ee5b2625b76a2e3f77b8ec1be7512cfb8d2cb) to not stale for 10s.

However nice side-effect your patch introduced - it made my test actually faster: `3.303s` vs `2.256s`. Thanks again!

**Comment by kenshaw** [comment]
I'm open to merging this; please rebase off the latest code and reduce to a single commit please.

(for the record, originally the approach we had taken was similar to this, but the implementation was changed for reasons I don't quite remember at the moment)

**Comment by stephencheng** [comment]
Hi @rjeczalik , was this issue resolved yet?

I am hitting the exactly same thing. Yes it's randomly waiting for a node for 7~10 seconds.

Thanks

**Comment by rjeczalik** [comment]
Oh my, sorry for spamming this issue with commit's references.

@stephencheng I'm using @pwaller's patches and works like a charm for me so far.

https://github.com/rjeczalik/chromedp

Maybe we should push forward with merging hanging PRs?

**Comment by stephencheng** [comment]
@rjeczalik I spent a fair bit of time to do two set of comparable tests.

For the Test A: a web site with lots of resources, such a linked js libs, iframes etc, the 7 seconds are roughly the time all resources being loaded. In this case, it seems the cdp is doing right and proper to wait till the end of everything.

For the Test B: same code, but use different site, eg: example.org, with no js, css etc. It took 2 seconds instead. So above so called bug does not exist at all.

I am not entirely sure if this is or should be a bug.

What I am not sure in Test A are:

why there is 7 seconds wait there? for nothing? 

```
	err = c.Run(ctxt, cdp.Tasks{
		cdp.Navigate(`https://a.web.site`),
})

 
2017/10/23 22:04:06 ---------------->start
2017/10/23 22:04:13 could not perform (attributeModified) operation on node 172 (wait node): timeout waiting for node `172`
2017/10/23 22:04:13 -> {"method":"Page.frameStartedLoading","params":{"frameId":"34805.9"}}

```

By tracing into the node 172 a little earlier, see below, it seems right before the navigation starts, it (cdp instance) was blocked to wait for chrome to load everything. Since our own code somehow uses cdp and underling resource in sort of async(go routine) proc, so it appears in log that it has wait/being hanging there for 7 seconds.

```
2017/10/23 22:04:03 -> {"method":"DOM.childNodeInserted","params":{"parentNodeId":169,"previousNodeId":171,"node":{"nodeId":172,"backendNodeId":113,"nodeType":1,"nodeName":"DIV","localName":"div","nodeValue":"","childNodeCount":0,"attributes":[]}}}
2017/10/23 22:04:03 -> {"method":"DOM.attributeModified","params":{"nodeId":172,"name":"id","value":"sizzle-1508756643942"}}
2017/10/23 22:04:03 -> {"method":"DOM.childNodeRemoved","params":{"parentNodeId":169,"nodeId":172}}
```

If so, should the wait be necessary? or the fix is to give an option that in such a case, it just do it in an unbloked way? @pwaller , if so, will it cause some other issue like our own code need to evaluate some result in js console, but since the resources were not loaded properly, it could result in unexpected/uncertain result?

PS. I used casperjs (underling is phantom js), it does not have such an issue of 7 seconds wait for resourced to finish loading.

**Comment by rjeczalik** [comment]
@stephencheng The reason why it happens is described in the issue title and @pwaller also commented with a more detailed description.

Long story short the extra seconds you observe comes from current event processing implementation, which is async and deals with issues by using timeouts. The issue here is the events can happen out of order, so if particular event is removed before other part of the code finishes waiting for it, the timeout kicks in and everything stales.

**Comment by rjeczalik** [comment]
> Trying to use your repo for a test but hitting the issue of below:

@stephencheng If you want to give it a try, you'd need to clone it under original import path.

```
$ git clone git@github.com:rjeczalik/chromedp src/github.com/knq/chromedp
```

**Comment by liudanking** [comment]
@rjeczalik It works like a charm for me too.

Hope it will be merged to master soon.

**Comment by stephencheng** [comment]
@rjeczalik @pwaller @kenshaw 

Attaching a testing report of comparison result with and without the fix for 1.  a normal case without the problem (using example.org) 2. a heavy complex site with the problem surfaced


- without the patch:

1. normal case

2017/10/23 21:54:28 normal case without fix loop(example.org) - 1 : 2.09315ms
2017/10/23 21:54:31 normal case without fix loop(example.org) - 2 : 1.736911ms
2017/10/23 21:54:34 normal case without fix loop(example.org) - 3 : 2.778175ms
2017/10/23 21:54:37 normal case without fix loop(example.org) - 4 : 3.153285ms
2017/10/23 21:54:40 normal case without fix loop(example.org) - 5 : 2.965242ms

2. problem surfaced:

2017/10/23 21:47:30 problematic case loop - 1 : 9.97912609s
2017/10/23 21:47:40 problematic case loop - 2 : 7.432463676s
2017/10/23 21:47:51 problematic case loop - 3 : 7.3978227s
2017/10/23 21:48:01 problematic case loop - 4 : 7.496083428s
2017/10/23 21:48:11 problematic case loop - 5 : 7.416036429s

- with the patch:

1. normal case

2017/10/23 21:40:40 fixed problematic case loop - 1 : 1.575ms
2017/10/23 21:40:43 fixed problematic case loop - 2 : 3.907262ms
2017/10/23 21:40:46 fixed problematic case loop - 3 : 2.569322ms
2017/10/23 21:40:49 fixed problematic case loop - 4 : 3.770594ms
2017/10/23 21:40:52 fixed problematic case loop - 5 : 3.452342ms

2. with problem surfaced:

2017/10/23 21:57:07 normal case (example.org) with fix loop - 1 : 1.898763ms
2017/10/23 21:57:10 normal case (example.org) with fix loop - 2 : 2.126858ms
2017/10/23 21:57:13 normal case (example.org) with fix loop - 3 : 1.771932ms
2017/10/23 21:57:16 normal case (example.org) with fix loop - 4 : 2.041133ms
2017/10/23 21:57:19 normal case (example.org) with fix loop - 5 : 2.073063ms

Conclusion:

The patch does not impact the normal case, but unblock the wait and increase the performance for about 5000 times.

>>> 10000/2.0
5000.0

I am hoping this patch could soon be merged back 

Thank you

**Comment by pwaller** [comment]
@stephencheng @liudanking and others, if the patches help, please click the [thumbs up on the PR](https://github.com/knq/chromedp/pull/77) to record that. I am not affiliated with the project, only a user myself.

**Comment by kenshaw** [comment]
Just want to give an update here: I will merge this stuff in, I just haven't had time to run it against other tests. Not trying to drag my feet on this, but have very little free time.

**Comment by pwaller** [comment]
No worries @kenshaw, thanks for the update.

**Comment by stephencheng** [comment]
@rjeczalik @kenshaw @pwaller 

I experienced a problem with random in consistence of grab data text, and I am not sure where the bug is or relating to applying @rjeczalik 's fix.

```
	err = c.Run(ctxt, cdp.Tasks{
		cdp.Click(`#arefreshlink`, cdp.NodeVisible),
		cdp.WaitVisible(`#alinkselector`),
		cdp.Text(sel1, &text1),
		cdp.Text(sel2, &text2),
		cdp.Text(sel3, &text3),
                .......
		cdp.Text(sel16, &text16),

	})

```

I make a for loop for 20 times, 1 in 5, or 20% of the chance is that 1 or 2 fields will be empty. However this is not the case for the web page, the page does show the text correctly.

To ensure that there is no data source error, I use casperjs to have tested same case and it never fail and actually I used it in production case, it just does not fail at all to get an empty text from chrome side.

So would this be a bug from: cdp.Text() func or this is due to the @rjeczalik 's patch? 

Have you got similar issue for you? Would you please have a look of this issue?

Thanks

**Comment by stephencheng** [comment]
@rjeczalik I just pulled all your latest code and used your patch, it worked and the above issue has disappeared. I still have one more scenario to test through, so I will update again on that.

**Comment by stephencheng** [comment]
@rjeczalik I think I have got what is wrong with the code. I will need to put a tiny sleep there for page elements to be fully loaded to ensure all texts has value.

So please ignore my report

**Comment by rjeczalik** [comment]
It sounds like there is a race somewhere, while a sleep might help I think
it is a fragile solution. You could give cdp.WaitLoad a try.

**Comment by stephencheng** [comment]
@rjeczalik I could not find WaitLoad from anywhere. where is it? why do you think that's the cause?

Anyway, I have done some extensive tests on this issue and I could summarise with this issue:
[https://github.com/knq/chromedp/issues/137](url)

**Comment by rjeczalik** [comment]
@stephencheng Sorry, it's cdp.WaitEventLoad - you use it after cdp.Nagivate etc. to ensure you start executing selectors after the page is fully loaded.

**Comment by stephencheng** [comment]
@rjeczalik this however does not help at all, due to the fact that the expected page elements not in those events. For example, it tool only 500 ms for those text element to show, however, it might take nearly 10 seconds for the page/events to be loaded. It's not relevant and can't help.

Thanks any way. The raised case is kind of bug as tested results show.

**Comment by rjeczalik** [comment]
> however, it might take nearly 10 seconds for the page/events to be loaded

So you need to wait for them separately with `WaitReady` / `WaitVisible` actions, e.g. find a common parent element for all of your text elements and wait for it to be visible.

**Comment by stephencheng** [comment]
@rjeczalik seems @kenshaw has split the codebase, have you got your patched codebase updated? It doesn't look like you have fetch the change and update yours from your repo.

What do I do if I need to port what your have patched previously?

@kenshaw appreciate if you can get this one merged sooner than later.

**Comment by swizzley** [comment]
So has anyone got this to work yet running in go routines? If I space out my routines by 200ms they fail immediately, 1 sec and the 2nd times out, I've yet to find a way to run routines any other way but serialized.

**Comment by stephencheng** [comment]
@swizzley you will have to stick to the old code(before codebase being split) and apply above mentioned fix. This is most people do and it's unofficial.

**Comment by ryderlee** [comment]
hi there. May I ask if there is a repo that I can use at the moment to avoid this issue? I tried @rjeczalik one but not success. thanks.

**Comment by trheyi** [comment]
It same to the problem is not be solved yet .   I try all the ways , but not success....

**Comment by mvdan** [comment]
Apologies all for the radio silence. I think this issue is, at core, very similar to #168. Please see [the comment I left there](https://github.com/chromedp/chromedp/issues/168#issuecomment-465550711) last month, and the progress I've made since.

@pwaller was right: it was wrong for events and commands to be handled out of order. The refactor branch fixes that, among other issues. The last comment I left in the other thread gives some details. It also asks for feedback - in particular, if anyone still encounters this bug in the refactor branch, please let me know.

**Comment by mvdan** [comment]
The refactor branch has been merged into master, and all these 10s timeouts are completely gone from the new codebase. Since I'm pretty sure that the bug is now fixed in master, I'm going to close it. If anyone still experiences the issue, or encounters any new bugs when using the latest master version, please provide details.
